File: .NET Core/ToyORB/Network/Requestor.cs
using System;
using System.IO;
using System.Net.Sockets;
using ToyORB.Messages;

namespace ToyORB.Network
{
    public class Requestor<TReq, TRes> where TReq : Message where TRes : Message
    {
        private readonly string _host;
        private readonly int _port;

        public Requestor(string host, int port)
        {
            _host = host ?? throw new ArgumentNullException(nameof(host));
            _port = port;
        }

        public TRes MakeRequest(TReq request)
        {
            using (var socket = new TcpClient(AddressFamily.InterNetwork))
            {
                socket.ConnectAsync(_host, _port).Wait();

                Console.WriteLine($"Writing request to socket {request}");
                var sendStream = new BinaryWriter(socket.GetStream());
                request.WriteInto(sendStream);
                sendStream.Flush();

                var recvStream = new BinaryReader(socket.GetStream());
                Message response = Message.ReadFrom(recvStream);
                Console.WriteLine($"Read response from socket {response}");

                return (TRes) response;
            }
        }
    }
}

File: Java/ToyORB/src/ToyORB/network/Requestor.java
package ToyORB.network;

import ToyORB.messages.Message;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.Socket;

public class Requestor<REQ extends Message, RES extends Message> {
	private final String host;
	private final int port;

	public Requestor(String host, int port) {
		Message.initReaders();
		this.host = host;
		this.port = port;
	}

	public RES makeRequest(REQ request) throws IOException {
		Socket socket = new Socket(host, port);

		System.out.println("Writing request to socket - " + request);
		DataOutputStream sendStream = new DataOutputStream(socket.getOutputStream());
		request.writeInto(sendStream);
		sendStream.flush();
		socket.shutdownOutput();

		DataInputStream recvStream = new DataInputStream(socket.getInputStream());
		RES response = Message.readFrom(recvStream);
		socket.shutdownInput();
		System.out.println("Read response from socket " + response);

		sendStream.close();
		recvStream.close();
		return response;
	}
}
